### PR TITLE
recursively set correct directory and file permission on webroot at t…

### PR DESCRIPTION
**Pull request by corbolais**
…he right time during installation

Signed-off-by: corbolais <corbolais@gmail.com>

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [ ] I have tested my proposed changes, and have included unit tests where possible.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)

---
**What does this PR aim to accomplish?:**
As detailed in issue #3175 (and apparently having occurred in #2195, too), there happen to be bad permissions of the `webroot` directory tree, the pihole-installation fails at the webroot stage. 

Having had a non-default `umask`, the resulting permissions on `${webroot}` lead to a _dysfunctional admin interface_.

**How does this PR accomplish the above?:**
Correctly set permissions recursively for `webroot` subdirectories and any file therein.

The removed original `chmod` only executed if `"${INSTALL_WEB_SERVER}" == true`. However, the webroot directory tree exists once `"${INSTALL_WEB_INTERFACE}" == true` is entered. So it seems sensible to move the statements above the original condition.

Now, why exactly the original chmod was where it was is unknown to me, this patch might be well intensioned but wrong nevertheless. Possible reason: the placement was intentional and not modifyable. In this case, I'm happy to rework this PR. Please explain the reasons.

There's more to it:
https://github.com/pi-hole/pi-hole/blob/782fec841ea85b29a8bf26023e615972de9d4717/automated%20install/basic-install.sh#L1932-L1935
In commit https://github.com/pi-hole/pi-hole/commit/f187b42a98b984f49d2df2ba20672f5f2497ceca the literal webroot path was refactored into `$webroot`, however, three months later commit https://github.com/pi-hole/pi-hole/commit/e19adccd9c623eeba20e46a6347108e917af133e regressed and reintroduced the literal webroot path, in addition to a move of chmod inside the condition (as already mentioned above). The regression looks like an oversight, the intention of the move of chmod is still unknown to me as the commit does not detail any reason behind the move but merely states the obvious.

_Furthermore, the permissions introduced are equally insufficient and do not account for a non-default umask._

As a result of the above, this PR moves the permission change again out of the condition _and_ enforces correct permissions _recursively_ on subdirectories and files of `webroot`. Only this warrants that
a) permissions are correct regardless of any umask setting
b) permissions are enforced from the moment the `webroot` directory exists.

It is debatable whether the parent directory of `$webroot` shall be included. Please advise on this.

Thank you.

**What documentation changes (if any) are needed to support this PR?:**
none.

### CODE REVIEW

**Comment by pvogt09** [comment]
The reintroduction of the hard coded path was requested by @Mcat12 in https://github.com/pi-hole/pi-hole/pull/2730#issuecomment-489372727 and https://github.com/pi-hole/pi-hole/pull/2730#discussion_r281032964, because he said that it is not allowed to change the path anyway and it will be removed soon.
The reason for introducing paths **above** `$webroot` was mainly that `www-data` might not have read or execute access to those directories and in turn is not able to read `$webroot` as far as I understood.
A possible solution for setting permissions for a variable `$webroot` up to `/` could be https://github.com/pi-hole/pi-hole/pull/2730#issuecomment-489370591.
The reason for using `a+rx` instead of a numerical value in that case was to keep other existing permissions for those directories untouched.
The files **below** `$webroot/admin` should be handled correctly by `getGitFiles` and `$webroot/pihole` by `installPiholeWeb` if I did not make a mistake in #2730 and that should also have fixed #2413 after 2e6198077d8f505415b8f494357eb40cdec12277 due to https://github.com/pi-hole/pi-hole/pull/2730#issuecomment-490735172.

It also might not be a good idea to change all files and directories in `$webroot` (not even with a+rx instead of 0775/0664) because there might already be other websites/virtual hosts located there with permissions set accordingly and changing everything to world readable might break other things, be a security risk or at be least too permissive.

@Mcat12: Probably, test cases for non default `umask`s could be added to the test suite in order to tackle those kinds of problems in the future.

@corbolais: Did you do a manual `git checkout/reset/...` on the webroot directory? That could also have changed the permissions when you did that with the changed `umask` which was the reason for me starting #2730 initially.

**Comment by Mcat12** [comment]
Thank you both for your analysis of the issue. Unfortunately the powers at be have decided to forgo some of the improvements which were expected to fix many of these underlying issues, and I have also largely moved on from the project. So I'll defer to @pi-hole/core-approvers.

**Comment by dschaper** [comment]
We thank mcat12 for his contributions, and wish him the best.

> As a result of the above, this PR moves the permission change again out of the condition and enforces correct permissions recursively on subdirectories and files of webroot. Only this warrants that
> a) permissions are correct regardless of any umask setting

Seems sane to do. `umask` is a preference of the user.

corbolais:
> b) permissions are enforced from the moment the webroot directory exists.

pvogt09: 
> It also might not be a good idea to change all files and directories in $webroot (not even with a+rx instead of 0775/0664) because there might already be other websites/virtual hosts located there with permissions set accordingly and changing everything to world readable might break other things, be a security risk or at be least too permissive.

Also sounds logical to me. But the argument about what permission and how forcefully it should be applied needs to be considered.

I'm fine with webroot being a variable as well. 

What is the end goal to be accomplished with this?

**Comment by corbolais** [comment]
@pvogt09 , @Mcat12 Thank you for coming back on this.

@pvogt09 Thank you for replying in such (differentiated) detail. I think I need to split this into separate answers to make handling this issue easier on the long run (divide et impera).

My impression is, given the length of and the level of detail already generated by this issue (not to even mentioning #2730), that there is a dire need to get 
a) this documented properly in the code as lengthy gh issues aren't a suitable drop-in replacement for explicit inline documentation
b) merged strictly in code only the separation (as of now) of semantics spread out into git repositories _and_ code. This is way to prone to fail and cumbersome to handle, prove is at hand. In this case the git repository should handle code, not semantics to be dealt with within "out-of-band" code, from the git repository's point of view.

> The reason for using `a+rx` instead of a numerical value in that case was to keep other existing permissions for those directories untouched.

ACK. I just wanted this to be dealt with and didn't care for the rest as I only have pihole using webroot.

> It also might not be a good idea to change all files and directories in $webroot (not even with a+rx instead of 0775/0664) because there might already be other websites/virtual hosts located there with permissions set accordingly and changing everything to world readable might break other things, be a security risk or at be least too permissive.

ACK. See above.

> @Mcat12: Probably, test cases for non default umasks could be added to the test suite in order to tackle those kinds of problems in the future.

ACK. This is mandatory, given the issue history these inconspicuous lines have generated already.

> @corbolais: Did you do a manual git checkout/reset/... on the webroot directory? That could also have changed the permissions when you did that with the changed umask which was the reason for me starting #2730 initially.

No, I did not intervene manually before the first error messages occurred (in the web interface), after the installation finished. I did however install pihole on an x86 host in this instance, that might be an important detail to mention at last.

**Comment by dschaper** [comment]
Okay, do you want to set up PRs to address the individual topics?

**Comment by corbolais** [comment]
> Okay, do you want to set up PRs to address the individual topics?

I think (read: hope) we might be able to get away with using this thread. So far.

**Comment by corbolais** [comment]
> Also sounds logical to me. But the argument about what permission and how forcefully it should be applied needs to be considered.

This is the main blocker for now: gather all the facts, read up on the history of the previous attempts to make it right and then hopefully in the end we should'nt be required to touch this any more. Until the next time.

> I'm fine with webroot being a variable as well.
ACK. This should definitely be/remain/become a variable. Consistently. Might require some code checking for occurences. Prepare for some getting overlooked.

> What is the end goal to be accomplished with this?
To never have to touch it again. ;-) I'll try to compile all conditions and set up a PR. There is definitely room for some major documentation improvement here. All the conditions and previous handlings of this part of the code need to be condensed into code documentation. And while we're at refactoring, the code might profit of a split of concern and responsibility. I'll give it a try. Might need a round of RFC before we got all requirements right. @pvogt09 already did a good job with the list of conditions he already provided us with.

**Comment by corbolais** [comment]
> because he said that it is not allowed to change the path anyway and it will be removed soon.

I challenge this. Anyone, please provide reasons for this path not to be available to changes. IMO, it might even be a better way to be able to change `webroot`.

**Comment by corbolais** [comment]
Conditions compiled so far are:

+ use symbolic, (partly) modifying permissions instead of numeric absolute, replacing permissions
+ warrant correct permissions in install script only. Do explicitely _not_ rely on external handling this reliably
+ introduce proper exhaustive documentation regarding this part of the install script
+ refactor the function, respect areas of concern, code responsability
+ unit test WIP https://github.com/pi-hole/pi-hole/pull/3191
+ recursive permission modification: requirements? objections? TBD

I shall update this list.

**Comment by corbolais** [comment]
> * webroot: how to handle parent dirs?

While technically, a `o+rx` on all path elements prior to `webroot` should be sufficient, a user/administrator might have objections regarding privacy. In this case a  `g+rx` together with a chmod .www-data might do.

RFC:
However, the second way might modify locally required permissions to be set. Here we leave well-trodden paths (no pun intended). This is a generic problem that requires a pihole-global decision on how to handle those cases. Another example of this is the lighttpd config file which in my case is not a default system-specific package-installed file (i.e. dpkg) but a locally modified one. Pihole up to now just replaces this file breaking existing configurations and installations. No good (tm).

Possible solutions are:
+ ask user
+ just do it (and possibly warn about it)
+ abort

**Comment by PromoFaux** [comment]
>And when is what (webroot?) supposed to be removed? For what reasons?

There _are_ plans on a new web interface and backend, but that's not planned until 6.0 (no ETAs - we haven't even got 5.0 out the door yet!)

At the time of that statement, I guess it was hoped that things would happen sooner, but the official stance at this time is.. "not for a while"

**Review by DL6ER — APPROVED** [review]
This seems good to go, however, there is one merge conflict currently blocking this

**Comment by MichaIng** [comment]
Just recognised the approval here despite the discussion does not seem to have a result?

IMO the PR is doing the opposite of what would be reasonable: Permissions should be touched as few as possible, only where necessarily required, while this PR forcefully and recursively applies permissive modes to directories which are not used at all by Pi-hole:
- Pi-hole only uses `/var/www/html/admin` and `/var/www/html/pihole`.
- `/var/www/html/admin` modes are applied via `git reset --hard` and even another `chmod -R a+rX` afterwards on fresh installs and update: https://github.com/pi-hole/pi-hole/blob/master/automated%20install/basic-install.sh#L632-L635
And: https://github.com/pi-hole/pi-hole/blob/master/automated%20install/basic-install.sh#L603-L615
- `/var/www/html/pihole` is as well installed with hard-coded permissions: https://github.com/pi-hole/pi-hole/blob/60fa93ab47e96fb99a864b3805572f209003383b/automated%20install/basic-install.sh#L1778-L1781
- Hence no further recursive permissions should be required at all.
- And there is no reason to grant group members write permissions to anything inside webroot. Not even `www-data` requires that, nor did it (or group members) have it on default Pi-hole installs. Not sure why it was granted on `${webroot}` (non-recursively!), there is no reason for that.

I just verified the above by adding `umask 0077` to the top if Pi-hole installer script, and all permissions inside webroot remain exactly the same as without it. Probably something has changed from the time on when this PR has been opened, but now it has no effects on any files used by Pi-hole but only might override permissions of files not used by Pi-hole which where intentionally set by the admin for other purpose.

What the initial code was aimed for and did was kinda the opposite: It is to assure that path elements from root until webroot are world-accessible (for `/` and `/var` is is assumed as otherwise ~no system would work at all), hence not from webroot down, but only up. It did this even more permissive than required. Since execute permissions only are sufficient to permit access to sub directories, enough would be:
```
chmod a+x /var/www /var/www/html
```

I was also a bid confused by the fact that `/var/www/html` as well as `${webroot]` is both used in the code, but I think it doesn't matter anymore now where Pi-hole v6 will ship with integrated webserver. So I would no bother with that at all and accept both `/var/www/html` and `${webroot}` as equal for as long as this code is relevant at all 😉. It would be only a matter of code beauty and consistency.

**Comment by corbolais** [comment]
About the hard coded path and webroot variable: it is important to come to terms consistently. So, who'd volunteer to replace those hard coded paths by the proper variables?

Generally, it's perferable to grant minimum working permissions. So the less bits set in the permission bitmask the better, that's why `umask 0077` was installed in the first place. Somehow PH install routines tripped on the umask thing and blew up ../admin webpages.

The PR sets the group write bit. As it's been PR'd in March I fail to recall whether I noticed some real need for this in my PH installation, i.e. it wasn't working without, or whether my particular larger setup requires this. In any case, I suspect it was essential to have it set the way the PR sets it.

If this PR gets dropped and what it fixed _for me_ (tm) happens once more I'll probably slam it in again. ;-)

cheers

**Comment by MichaIng** [comment]
I just checked where the hardcoded webroot path has been added. It was actually here: #2730
I guess it was simply overseen that `${webroot}` should have been used instead. So looks like it can be changed without breaking anyone's aim.

I remember some discussion around permissions and whether they should be changed when someone chooses to not install Lighttpd via installer, and that the result was that users who do so will probably have a custom setup that should not be messed with and to hence not set any permissions then. But AFAIR it was about `usermod -a -G pihole ${LIGHTTPD_USER}` a few lines below (`usermod -a -G ${LIGHTTPD_GROUP} pihole` has already been removed as it is not required). But at least execute permissions to grant any webserver user "navigation" access to a sub directory within is so essential that IMO it makes sense to apply it regardless of webserver choice. With the group membership instead it could be argued that a custom webserver might run as different user (i.e. not `www-data`).

Finally my suggestion for this block of code would be:
```
    if [[ "${INSTALL_WEB_INTERFACE}" == true ]]; then
        if [[ ! -d "${webroot}" ]]; then
            # make the Web directory if necessary
            install -d -m 0755 ${webroot}
        else
            # else assure that any webserver have sub directory access
            chown a+x ${webroot}
        fi

        # Assure that any webserver has access to webroot and sub directories, which requires execute permissions on the parent directory
        chown a+x /var/www

        if [[ "${INSTALL_WEB_SERVER}" == true ]]; then
            # Add lighttpd to the pihole group so the web interface can manage the gravity.db database
            usermod -a -G pihole ${LIGHTTPD_USER}
```
It still looks a bit inconsistent as `/var/www` again is hard-coded. `${webroot%/*}` could be used to apply it to the parent directory, regardless which path it is, but in case `${webroot} == '/something'` this would be an empty string 😉.

**Comment by corbolais** [comment]
> I just checked where the hardcoded webroot path has been added. It was actually here: #2730
> I guess it was simply overseen that ${webroot} should have been used instead.

This was my best guess as well. I didn't delve deeper nor did I try to reason why and when as I wasn't involved back then.

>I remember some discussion around permissions and whether they should be changed when someone chooses to not install Lighttpd via installer, and that the result was that users who do so will probably have a custom setup that should not be messed with and to hence not set any permissions then. But 

ACK

> But at least execute permissions to grant any webserver user "navigation" access to a sub directory within is so essential that IMO it makes sense to apply it regardless of webserver choice. 

ACK

> With the group membership instead it could be argued that a custom webserver might run as different user (i.e. not www-data).

ACK

I still feel that nagging sensation I am missing the point why I set permissions as I did. However, I couldn't be bothered to check and verify the running setup right now. There must have been "a good reason" at the time, surely. I'd be rather interesting to see whether it will fail again with a fresh installation in this particular setup. ;-)

Well, `"${webroot%/*}"` is greedy. It will reduce anything resembling a path like `"webroot=/foobar"` to `''`, regardless. Either use `"${webroot/\/*/\/}"` or guard it using `defaultPath="/"; .. "${webroot:=defaultPath}" ..`.

It would be sensible, though, to use `dirname "$webroot"`, anyway.

cheers

**Comment by MichaIng** [comment]
> Well, "${webroot%/*}" is greedy.

It is not, `"${webroot%%/*}"` would be greedy. But yes, as said, it would turn a single path element into an empty string.

> dirname "$webroot"

Good alternative, actually, same result but does not suffer from the empty string issue (instead correctly returns `/` as last result). Would be nice to have something like `chmod -p` (p=parent) as opposite of `chmod -r` to apply things to all parents up to root, similar to `rmdir -p` or `mkdir -p`, so no directory tree climbing would be required 😄.

> There must have been "a good reason" at the time, surely. I'd be rather interesting to see whether it will fail again with a fresh installation in this particular setup. ;-)

Probably there was a reason when you opened the PR, which is nearly a year ago. Of course it would still be great if you could test it, probably I have overseen something.
____________
Btw I found #3191 and like the idea to create unit tests for exactly such issues. Do I see it right that you basically created a whole installer test for this? Great for so many other reason, but when checking webroot permissions, be sure to not test the permission bits of each file/directory, but instead the result serves the needs, i.e. that the webserver user can read all files in `/var/www/html/(admin|pihole)`, if that is possible, so checking the parent directories explicitly could be skipped. Otherwise, as said above, the parent directories only need to be executable, not readable, to allow sub directory access.

**Comment by corbolais** [comment]
> > Well, "${webroot%/*}" is greedy.
> 
> It is not, `"${webroot%%/*}"` would be greedy

ACK. Must have been absent-minded while writing this.

> Btw I found #3191 and like the idea to create unit tests for exactly such issues. Do I see it right that you basically created a whole installer test for this?

True, but credits @pvogt09 .

**Comment by MichaIng** [comment]
By chance I just saw that parent path execute permissions loop being used in the oldest open PR here: https://github.com/pi-hole/pi-hole/pull/3016/files#diff-595630a29a855f6d667a84ca0662042e826bf3ec56322ef61d4a6ef149147d23R1930-R1934
`chmod a+x` again would be sufficient, but such loop is the most consequent step when worrying about missing access permissions.

As we discussed `$webroot` vs hard-coded paths, the above PR would create some consistency as well Pi-hole paths actually do have variables, but those are nur used everywhere.

**Comment by PromoFaux** [comment]
Closing due to no activity for nearly a year now. We can reopen/revisit if there is any interest, but things have probably changed so much since then that it will likely be better to start again

**Pull request closed without merging**
